### PR TITLE
Convert all remaining templates to HTML5

### PR DESCRIPTION
**Pull request by andrepereiradasilva**
Pull Request for Improvement.
#### Summary of Changes

Basic conversion of all site and admin templates to HTML5 (some where still XHTML 1.0 Transitional).
#### Testing Instructions
1. Install patch
2. Check if pages in all templates (normal, component popup - ex: mail/print, error pages, offline and login) site and admin are working properly.

Check also the code changes.
#### Observations

All should now start with this code:

``` html
<!DOCTYPE html>
<html lang="en-gb" dir="ltr">
<head>
[...]
    <meta charset="utf-8" />
[...]
```

Suggestions and/or improvements are welcomed.

### CODE REVIEW

**Comment by Bakual** [comment]
Are beez and hathor really HTMl5 templates? I don't know myself.

**Comment by andrepereiradasilva** [comment]
#### beez3

just check the generated HTML code of the first page. You can see footer, section, tags that are exclusive HTML5. So this really fixes a bug.

Also checked the frontpage generated HTML code after this PR in the [W3C validator](https://validator.w3.org/) and the only errors i got are regarding the CSS `media=projection` that is deprecated, I plan to remove those too in this PR.

#### hathor

I think it was not build as an HTML5 template, but in the generate code across the backend exist some HTML5 features (`data- *` attributes for instance) so it should be HTML5.

Also checked the control panel generated HTML code after this PR in the [W3C validator](https://validator.w3.org/) and just got 1 error: `align="center"` (in the version) that is obsolete in HTML5. I plan to remove that too in this PR.

I also plan to put html5.js for older IE versions.

**Comment by brianteeman** [comment]
> Are beez and hathor really HTML5 templates? I don't know myself.
For beez at least
https://issues.joomla.org/tracker/joomla-cms/7359#event-139164

File: administrator/templates/hathor/error.php [hunk 1]
```diff
@@ -25,7 +28,7 @@
 	<?php if ($this->direction == 'rtl') : ?>
 		<link href="<?php echo $this->baseurl; ?>/templates/<?php echo  $this->template; ?>/css/template_rtl.css" rel="stylesheet" type="text/css" />
 	<?php endif; ?>
-
+	<!--[if lt IE 9]><script src="<?php echo $this->baseurl; ?>/media/jui/js/html5.js"></script><![endif]-->
```
Comment (brianteeman) [inline administrator/templates/hathor/error.php]: Why is this removed?

File: templates/beez3/component.php [hunk 2]
```diff
@@ -10,12 +10,16 @@
...
+// Output as HTML5
+$doc->setHtml5(true);
+
 $color = $this->params->get('templatecolor');
 
 $doc->addStyleSheet($this->baseurl . '/templates/system/css/system.css');
-$doc->addStyleSheet($this->baseurl . '/templates/' . $this->template . '/css/template.css', $type = 'text/css', $media = 'screen,projection');
-$doc->addStyleSheet($this->baseurl . '/templates/' . $this->template . '/css/position.css', $type = 'text/css', $media = 'screen,projection');
-$doc->addStyleSheet($this->baseurl . '/templates/' . $this->template . '/css/layout.css', $type = 'text/css', $media = 'screen,projection');
+$doc->addStyleSheet($this->baseurl . '/templates/' . $this->template . '/css/template.css', $type = 'text/css', $media = 'screen');
```
Comment (brianteeman) [inline templates/beez3/component.php]: why is projection removed?

**Comment by brianteeman** [comment]
Other than reducing th file size by a few bytes is there any benefit to these changes

File: templates/beez3/component.php (hunk 2 above)
Comment (andrepereiradasilva) [inline templates/beez3/component.php]: because the W3C validator said that media=projection is deprecated in html5.

File: administrator/templates/hathor/error.php (hunk 1 above)
Comment (andrepereiradasilva) [inline administrator/templates/hathor/error.php]: removed? it was added a missing html5.js

File: templates/beez3/component.php (hunk 2 above)
Comment (brianteeman) [inline templates/beez3/component.php]: OK thanks - found this link http://stackoverflow.com/questions/31410984/link-tag-media-attribute-projection-tv-not-valid so that satisfies my enquiry

File: administrator/templates/hathor/error.php (hunk 1 above)
Comment (brianteeman) [inline administrator/templates/hathor/error.php]: I need more coffee

**Comment by andrepereiradasilva** [comment]
> Other than reducing the file size by a few bytes is there any benefit to these changes

The few bytes reduced are not relevant to the case.

The benefits? Aside that with the creation of HTML5. XHTML is good for the history books? :)

It solves a series of problems of using HTML5 features (section, footer, header HTML tags, data-* attributes, etc) in the Joomla core that are rendered in this templates inside a XHTML 1.0 Transitional DOCTYPE which in invalid HTML code according to the web standards.

You can check that in the W3C validator: https://validator.w3.org/

Example for beez3:
image

Also, one of the things accessible templates need is having valid HTML code so that machines can correctly interpret the HTML code.

**Comment by brianteeman** [comment]
Thank you that answers my question

**Comment by andrepereiradasilva** [comment]
ok, it's ready for testing.

**Comment by franz-wohlkoenig** [comment]
I have tested this item :white_check_mark: successfully on 920bb31cb7e520977d11ef60684b1f69214840e6<br /><br />Working properly & start with correct code:

**Isis**
- System: Control Panel, Global Configuration, Global Check-in, Clear Cache, Clear Expired Cache.
- Component: Banners, Categories, Clients, Tracks.
- Error Page

**Hathor**
- System: Control Panel, Global Configuration, System, Server, Permissions, Text Filters.
- Component: Banners, Categories, Clients, Tracks.
- Error Page

**Protostar**
- Default Startsite
- Error Page

**Beez3**
- Default Startsite
- Error Page<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9842">issues.joomla.org/joomla-cms/9842</a>.</sub>

**Comment by waader** [comment]
I have tested this item :white_check_mark: successfully on a231b51d64e494b1949e4a4810cc7cd6daf5a476<br /><br />Thanks!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9842">issues.joomla.org/joomla-cms/9842</a>.</sub>

**Comment by klatte88** [comment]
I found an issue, when swapping from protostar to beez template my article which was set to acces: registered with a readmore link. 

What is expected is that the intro shows with a readmore link that asks you to register/login
When swapping to beez from Protostar, the readmore link is gone and the full article and full image displays. In the backend the article still has a readmore links set<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9842">issues.joomla.org/joomla-cms/9842</a>.</sub>

**Comment by klatte88** [comment]
I have tested this item :red_circle: unsuccessfully on a231b51d64e494b1949e4a4810cc7cd6daf5a476<br /><br />changing from protostar to beez template breaks readmore link<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9842">issues.joomla.org/joomla-cms/9842</a>.</sub>

**Comment by klatte88** [comment]
Just to add if you want to see the issue I found. 

create an article following the steps described here (with or without patch both give the same issue):
https://issues.joomla.org/tracker/joomla-cms/9865<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9842">issues.joomla.org/joomla-cms/9842</a>.</sub>

**Comment by waader** [comment]
@klatte88 "with or without patch both give the same issue"

So the issue you found is unrelated to this patch, right?

**Comment by klatte88** [comment]
What I ment is with or without Brians patch for the other issue, to show that it is unrelated to Brians patch, I guess that was a bit vague<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9842">issues.joomla.org/joomla-cms/9842</a>.</sub>

**Comment by klatte88** [comment]
But yes upon further testing it seems to also be unrelated to your patch<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9842">issues.joomla.org/joomla-cms/9842</a>.</sub>

**Comment by waader** [comment]
@klatte88 
Thanks for clarifications! Please remove your unsuccessful test statement.

**Comment by mahagr** [comment]
@andrepereiradasilva 

Unfortunately this pull request (as it is) breaks some error pages. Just a week ago I fixed a major bug in Gantry 4, which was caused by the same behaviour.

You can quickly get the error by appending `&format=htm` to almost any page to get:

    Fatal Error: Call to undefined method JDocumentRaw::setHtml5()

Error page does not always have `JDocumentHtml` object, which causes $doc->setHtml5() to fail on fatal error.

Easiest fix is to add a check to make sure the function exists or the object is `JDocumentHtml` before attempting to make the call.

Proper fix would be to make it possible to change document type before rendering error page. Right now doing that is impossible, though.

**Comment by andrepereiradasilva** [comment]
thanks for the warning @mahagr. will check that

@mbabker can you give a hand here? is there is any reason why the error pages (error.php) are not rendered like index, login, component and offline templates?

That cause several issues and makes the job harder to make a custom templating to error pages.

**Comment by mbabker** [comment]
Ask whoever wrote `JDocumentError` to not inherit `JDocumentHtml` properties and functions.  error.php is rendered by the former, the rest by the latter.

**Comment by mahagr** [comment]
Actually `JDocumentError` is not being used either in this case; it uses `JDocumentRaw` which is set on routing (`htm` isn't a known type).

**Comment by andrepereiradasilva** [comment]
how about replacing (only in `error.php` files) 

```php
$doc = JFactory::getDocument();
```

for

```php
$doc = JDocument::getInstance('html');
```

To force the error pages to be rendered has HTML.

**Comment by mbabker** [comment]
That won't force it because it's a different document type.  You need to do
all the conversions in the error.php template files or make error.php be
rendered by the JDocumentHtml class.

**Comment by andrepereiradasilva** [comment]
> That won't force it because it's a different document type.

The change i propose in `error.php` files would solve the problem @mahagr is talking. I tested.

> or make error.php be rendered by the JDocumentHtml class.

Yes, IMHO `error.php` should be rendered by the JDocumentHtml class. That way it would be a template like the other ones (login, index, offline, component) with no special cases and so easier to maintain.

It would also solve some other bugs in error.php rendering (the mod_languages, for instance, doesn't load the css in error php pages).
image

But, since i don't understand quite well joomla error handling, if i try to do that kind of change the result will most probably be that something gets broken :smile: .

**Comment by mbabker** [comment]
Again, ask whoever designed `JDocumentError` to be a different document type than `JDocumentHtml`, and why they wrote the error handler to only support HTML output.  It's not a bug in the error.php template, it's how the thing is (poorly) designed; it doesn't support `<jdoc>` tags.  Changing the `$doc` reference won't fix anything in error.php; now you're changing references of another `JDocument` instance instead of `$this` instance that you're already rendering.  So you have to do all the HTML5 stuff in error.php directly and not rely on anything that's in the `JDocumentHtml` API.

**Comment by andrepereiradasilva** [comment]
ok, i see. since it was `(poorly) designed` this way i will do the small html5 changes in error.php directly then.

**Comment by andrepereiradasilva** [comment]
And thanks again for the explanations.

**Comment by andrepereiradasilva** [comment]
Pplease test again.
Include also a test in each template error page with `?format=raw` query string.

**Comment by mahagr** [comment]
After a quick code review, all error pages should now work.

**Comment by andrepereiradasilva** [comment]
This PR get conflicts really quickly.
So will close this PR and will make a PR for each template. That should be more easy to test and mantain conflicts.
